Unit testing with copilot instructions

Burcu-Ozkan opened the pull request.

Burcu-Ozkan committed:
Refactor testing structure and update documentation

- Updated `copilot-instructions.md` with detailed project architecture, unit testing rules, and examples.
- Removed `GoalManager.UnitTests` project; added `GoalManager.UseCases.Tests` project in the solution.
- Changed `AddGoalCommandHandler` from internal to public and added a resilience pipeline for concurrency handling.
- Removed `Microsoft.AspNetCore.Mvc.TagHelpers` from `_ManageNav.cshtml` and `_LoginPartial.cshtml`.
- Cleaned up `GlobalUsings.cs` by removing several global using directives, including `Xunit`.
- Updated `GoalManager.IntegrationTests.csproj` to include `NSubstitute` and remove `GoalManager.UnitTests`.
- Removed `NoOpMediator.cs` and `xunit.runner.json` files, simplifying the test configuration.
- Added/updated tests in `AddGoalCommandHandlerFixture.cs`, `AddGoalPeriodCommandHandlerTests.cs`, and `ApproveGoalProgressCommandHandlerTests.cs` to enhance command handler validation.
- Reorganized dependencies in `GoalManager.UseCases.Tests.csproj`.

Burcu-Ozkan committed:
Fix Turkish characters and improve clarity in docs

Updated `copilot-instructions.md` to correct Turkish language characters and enhance text clarity. Sections were revised for proper spelling and grammar, especially regarding software architecture and unit testing guidelines. Some content was reformatted for consistency, examples were clarified, and the checklist for new test writing was expanded for thoroughness.

QuickOrBeDead commented on ddd/goal-management-system/tests/GoalManager.UseCases.Tests/GoalManagement/AddGoal/AddGoalCommandHandlerTests.cs:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  [Fact]
+  public async Task Handle_Retries_on_initial_Concurrency_then_succeeds()
```
@Burcu-Ozkan  bu testi copilot mı yazdı?

QuickOrBeDead commented on ddd/goal-management-system/tests/GoalManager.UseCases.Tests/GoalManagement/AddGoal/AddGoalCommandHandlerFixture.cs:
```diff
@@ -0,0 +1,219 @@
+    return gs;
+  }
+
+  private static void SetProp(object target, string name, object? value)
```
@Burcu-Ozkan  reflection kullanmadan yapılamıyor mu?

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead commented on ddd/goal-management-system/tests/GoalManager.UseCases.Tests/GoalManagement/AddGoalPeriod/AddGoalPeriodCommandHandlerTests.cs:
```diff
@@ -0,0 +1,85 @@
+    await goalPeriodRepository.Received(1).AddAsync(Arg.Is<GoalPeriod>(p => GetProp<int>(p,(nameof(GoalPeriod.TeamId))) == cmd.TeamId && GetProp<int>(p, nameof(GoalPeriod.Year)) == cmd.Year), Arg.Any<CancellationToken>());
+  }
+
+  private static void SetId(object entity, int id)
```
@Burcu-Ozkan reflection kullanmadan yapılamıyor mu?

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead commented on ddd/goal-management-system/tests/GoalManager.UseCases.Tests/GoalManagement/AddGoalPeriod/AddGoalPeriodCommandHandlerTests.cs:
```diff
@@ -0,0 +1,85 @@
+    }
+  }
+
+  private static T GetProp<T>(object entity, string name)
```
@Burcu-Ozkan  reflection kullanmadan yapılamıyor mu?

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead commented on ddd/goal-management-system/tests/GoalManager.UseCases.Tests/GoalManagement/ApproveGoalProgress/ApproveGoalProgressCommandHandlerTests.cs:
```diff
@@ -0,0 +1,106 @@
+    await goalSetRepository.Received(1).UpdateAsync(goalSet, Arg.Any<CancellationToken>());
+  }
+
+  private static void SetId(object entity, int id)
```
@Burcu-Ozkan reflection kullanmadan yapılamıyor mu?

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead commented on ddd/goal-management-system/src/GoalManager.UseCases/GoalManagement/AddGoal/AddGoalCommandHandler.cs:
```diff
@@ -7,7 +7,7 @@
 
 namespace GoalManager.UseCases.GoalManagement.AddGoal;
 
-internal sealed class AddGoalCommandHandler(IRepository<GoalSet> goalSetRepository) : ICommandHandler<AddGoalCommand, Result<(int TeamId, int PeriodId)>>
```
@Burcu-Ozkan  internal'a test yazılabilir public'e gerek yok

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead commented:
bütün use case'lere test yazabiliyor mu?

QuickOrBeDead commented on ddd/goal-management-system/src/GoalManager.UseCases/GoalManagement/AddGoal/AddGoalCommandHandler.cs:
@Burcu-Ozkan internal'a test yazılabilir public'e gerek yok

Burcu-Ozkan replied:
kontrol edebilirsin

QuickOrBeDead reviewed (commented):
finito

Burcu-Ozkan committed:
Improve tests and refactor Goal Management handlers

- Added and modified unit tests for better coverage and edge case handling.
- Refactored methods to reduce unnecessary abstraction and improve readability.
- Enhanced error messages and assertions for clearer feedback.
- Introduced new test cases for recently implemented features.
- Updated access modifiers for better encapsulation and best practices.
- Added necessary using directives and assembly attributes for maintainability.

Burcu-Ozkan commented:
> bütün use case'lere test yazabiliyor mu?

yazdı, kontrol edebilirsin
